gcloud/datastore/query.py

```diff
@@ -104,7 +102,7 @@ def dataset_id(self):
 
         :rtype: str
         """
-        return self._dataset_id
+        return self._dataset_id or self._client.dataset_id
```

Comment:
Would it make more sense to do this in the constructor so we don't have a branch every time the property is accessed?